### PR TITLE
MINOR: Factor `RaftManager` out of `TestRaftServer`

### PR DESCRIPTION
**Pull request by hachikuji**
This patch factors out a `RaftManager` class from `TestRaftServer` which will be needed when we integrate this layer into the server. This class encapsulates the logic to build `KafkaRaftClient` as well as its IO thread.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/raft/KafkaNetworkChannel.scala [hunk 1]
```diff
@@ -87,6 +87,7 @@ private[raft] class RaftSendThread(
 
   def sendRequest(request: RequestAndCompletionHandler): Unit = {
     queue.add(request)
+    wakeup()
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/KafkaNetworkChannel.scala:90]: Is this a bug fix?

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  private val raftClient = buildRaftClient()
+  private val raftIoThread = new RaftIoThread(raftClient)
+
+  def startup(): Unit = {
+    netChannel.start()
+    raftClient.initialize()
+    raftIoThread.start()
+  }
+
+  def shutdown(): Unit = {
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala:126]: Should we shutdown ```raftClient``` in this method?

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    listener: RaftClient.Listener[T]
+  ): Unit = {
+    raftClient.register(listener)
+  }
+
+  override def scheduleAppend(
+    epoch: Int,
+    records: Seq[T]
+  ): Option[Long] = {
+    val offset: java.lang.Long = raftClient.scheduleAppend(epoch, records.asJava)
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala:144]: (this is unrelated to this PR.)

How about using ```Optional``` to replace the ```null```?

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+          }
+        }
+        true
+      } else {
+        false
+      }
+    }
+
+    override def isRunning: Boolean = {
+      client.isRunning
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala]: Should it consider the state of "this thread also"? for example:
```scala 
  super.isRunning && client.isRunning
```

File: core/src/main/scala/kafka/tools/TestRaftServer.scala [hunk 5]
```diff
@@ -131,162 +102,29 @@ class TestRaftServer(
...
 
-    socketServer.startProcessingRequests(Map.empty)
-    ioThread = new RaftIoThread(raftClient)
-    ioThread.start()
     workloadGenerator.start()
+    raftManager.startup()
+    socketServer.startProcessingRequests(Map.empty)
   }
 
   def shutdown(): Unit = {
```
Comment (chia7712) [inline core/src/main/scala/kafka/tools/TestRaftServer.scala:110]: Is it need to shutdown ```raftManager```?

**Review by chia7712 — COMMENTED** [review]
@hachikuji Thanks for nice refactor. Please take a look at following comments.

File: core/src/main/scala/kafka/raft/KafkaNetworkChannel.scala (hunk 1 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/KafkaNetworkChannel.scala:90]: Yes indeed, thanks for noticing. I guess your point is that we should have a test case. Let me see about writing one.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala:144]: Yeah, it's a good idea. I used the `null` as a placeholder until I thought of a better idea, but `Optional` is probably just as good as a placeholder and safer.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala:126]: See the logic in `RaftIoThread.initiateShutdown`. Note that the thread continues polling until graceful shutdown has either completed or timed out.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 4 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala]: I think not. `super.isRunning` will return true as soon as `initiateShutdown` is called, but I want the thread to keep polling until the client itself says it is time to shutdown.

Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala]: I've added a test case which hopefully makes the expected behavior clearer. Let me know if it seems reasonable.

File: clients/src/test/java/org/apache/kafka/clients/MockClient.java [hunk 6]
```diff
@@ -263,8 +263,11 @@ private synchronized void maybeAwaitWakeup() {
             if (remainingBlockingWakeups <= 0)
                 return;
 
-            while (numBlockingWakeups == remainingBlockingWakeups)
-                wait();
+            TestUtils.waitForCondition(() -> {
+                if (numBlockingWakeups == remainingBlockingWakeups)
+                    MockClient.this.wait(500);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/MockClient.java:268]: ```TestUtils.waitForCondition``` already call ```sleep``` so is this ```wait``` necessary?

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+object KafkaRaftManager {
+  class RaftIoThread(
+    client: KafkaRaftClient[_]
+  ) extends ShutdownableThread(
+    name = "raft-io-thread",
+    isInterruptible = false
+  ) {
+    override def doWork(): Unit = {
+      client.poll()
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala:49]: just curious. If ```poll``` throws exception, this I/O thread get died and ```isRunning``` still returns ```true``` (since raft client is running). How ```KafkaRaftManager``` deal with that scenario?

**Review by chia7712 — APPROVED** [review]
@hachikuji Thanks for nice refactor and test cases. just left two trivial comments. +1 to merge.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala:144]: I started to implement this, but found it was taking me a bit afield. If you don't mind, let's address this in https://issues.apache.org/jira/browse/KAFKA-12158.

File: clients/src/test/java/org/apache/kafka/clients/MockClient.java (hunk 6 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/MockClient.java:268]: I was thinking the wait would allow for faster notification.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 7 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala:49]: Good point. I changed `isRunning` so that it also checks if the thread has failed. Added a test case as well.

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+          }
+        }
+        true
+      } else {
+        false
+      }
+    }
+
+    override def isRunning: Boolean = {
+      client.isRunning && !isThreadFailed
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala:68]: There is another issue. After ```client.poll()``` gets failed, the ```isThreadFailed``` becomes ```true``` and so ```isRunning``` returns ```false```. It means ```initiateShutdown``` does not call ```client.shutdown``` even though we call ```shutdown``` explicitly.

**Review by chia7712 — COMMENTED** [review]
@hachikuji Thanks for updating code. There is another question. Please take a look.

File: core/src/main/scala/kafka/raft/RaftManager.scala (hunk 8 above)
Comment (hachikuji) [reply core/src/main/scala/kafka/raft/RaftManager.scala:68]: It's a good question and I was debating it earlier today. It's worth keeping in mind that `KafkaRaftClient` doesn't really own the resources that would need closing. Sockets and files are owned by `RaftManager`, so they will still get closed properly. The one exception is de-registration of metrics in `KafkaRaftMetrics.close`. 

The main purpose of `shutdown` is to attempt a graceful departure from the quorum. But after we after we have hit an unexpected exception and are left in an undetermined state, do we still want to try the graceful shutdown? I was leaning toward 'no'. 

Perhaps what this suggests is that we should have a separate `close` method which can handle cleanup of any resources that `KafkaRaftClient` owns today (and might own in the future). If we hit an uncaught exception, then we can forego the call to `shutdown`, but we can ensure that `close` always gets called. How does that sound?

Comment (chia7712) [reply core/src/main/scala/kafka/raft/RaftManager.scala:68]: > we can forego the call to shutdown, but we can ensure that close always gets called. How does that sound?

sounds good to me :)

**Comment by hachikuji** [comment]
@chia7712 Thanks for reviews. Take one more look at the updates when you have time.

**Review by chia7712 — APPROVED** [review]
@hachikuji Thanks for this great refactor. LGTM
